File: java/src/main/java/it/giannibombelli/refactor_challenge/FizzBuzz.java
package it.giannibombelli.refactor_challenge;

import java.util.List;

public class FizzBuzz implements FizzBuzzInterface {
    private final List<DivisorAndWord> divisorAndWords;

    public FizzBuzz(List<DivisorAndWord> divisorAndWords) {
        this.divisorAndWords = divisorAndWords;
    }

    @Override
    public String say(int number) {
        var result = "";
        for (DivisorAndWord divisorAndWord : divisorAndWords) {
            result += divisorAndWord.getWordOrEmpty(number);
        }

        if (!result.equals("")) {
            return result;
        }

        return String.valueOf(number);
    }
}

File: kotlin/src/main/kotlin/it/giannibombelli/refactor_challenge/FizzBuzz.kt
package it.giannibombelli.refactor_challenge

class FizzBuzz : FizzBuzzInterface {
    override fun say(number: Int): String {
        val pair = Pair(3, "Fizz")
        val divisorAndWordList = listOf(pair, Pair(5, "Buzz"))

        var result = ""

        for (item in divisorAndWordList) {
            result += say(number, item)
        }

        if (result.isNotEmpty())
            return result

        return number.toString()
    }

    private fun say(number: Int, item: Pair<Int, String>): String {
        return if (number % item.first == 0) {
            item.second
        } else {
            ""
        }
    }
}
